js/components/generate/GenerateForm.js

```diff
@@ -230,35 +235,56 @@ export default class GenerateForm extends React.Component {
       })
   }
 
-  /**
-   * Produce an appropriate warning message string in the case that no courses have been entered
-   * or that one or more invalid courses have been entered.
-   * @param {string[]} invalidCourses - The array of invalid course codes
-   * @returns {string} The warning message string.
-   */
-  computeMessage(invalidCourses) {
-    if (invalidCourses.length === 0) {
-      return "Cannot generate graph – no courses entered!"
-    } else if (invalidCourses.length === 1) {
-      return `The course ${invalidCourses} was invalid! Please check your input.`
+  validateForm = values => {
+    const errors = {}
+
+    const coursePattern = /^[A-Z]{3}\d{3}[HY]\d$/
+    const deptPattern = /^[A-Z]{3}$/
+
+    if (!values.courses.trim().length) {
+      errors.courses = "Cannot generate graph – no courses entered!"
     } else {
-      return `The courses [${invalidCourses}] were invalid! Please check your input.`
+      const courses = values.courses.split(",").map(course => course.trim())
+      const invalidCourses = courses.filter(course => !coursePattern.test(course))
+
+      if (invalidCourses.length > 0) {
+        errors.courses =
+          invalidCourses.length === 1
+            ? `The course ${invalidCourses} was invalid! Please check your input.`
+            : `The courses [${invalidCourses}] were invalid! Please check your input.`
```

Comment:
Please use `Array.prototype.join` here, so that there's a space after each comma. Same for all of the error messages involving lists.